js/modules/k6/data/share.go

```diff
@@ -112,3 +112,18 @@ func (s wrappedSharedArray) deepFreeze(rt *goja.Runtime, val goja.Value) error {
 	}
 	return nil
 }
+
+// SharedArrayBuffer holds the bytes of a shared array buffer.
+type SharedArrayBuffer struct {
+	arr []byte
+}
+
+// Wrap wraped a SharedArray buffer with a provided JS runtime
+func (sab SharedArrayBuffer) Wrap(runtime *goja.Runtime) goja.ArrayBuffer {
+	// return rt.NewArrayBuffer(sab.arr)
+	// return rt.NewDynamicArray(WrappedSharedArrayBuffer{
+	// 	SharedArrayBuffer: sab,
+	// 	rt:                rt,
+	// })
+	return runtime.NewArrayBuffer(sab.arr)
```

Comment:
This in no way prevents the underlying `[]byte` from being changes as shown by this 
```javascript
import { sleep } from "k6";
var b = open("./sab.js", "rb")

export default function() {
	var v = new Int8Array(b);
	v[0] = __VU + __ITER
	sleep(1)
	console.log(v[0]);
}
```
(name it `sab.js` for easier)
running with `go run -race . run -u 20 - i 30 sab.js` I get 
```
WARNING: DATA RACE
Write at 0x00c000546840 by goroutine 45:
  github.com/dop251/goja.(*int8Array).set()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/typedarrays.go:198 +0x7e
  github.com/dop251/goja.(*typedArrayObject)._putIdx()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/typedarrays.go:520 +0x108
  github.com/dop251/goja.(*typedArrayObject).setOwnIdx()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/typedarrays.go:542 +0x4a
  github.com/dop251/goja.(*Object).setOwn()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/object.go:1403 +0x16c
  github.com/dop251/goja._setElemStrictP.exec()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/vm.go:1403 +0x235
  github.com/dop251/goja.(*_setElemStrictP).exec()
      <autogenerated>:1 +0x44
  github.com/dop251/goja.(*vm).run()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/vm.go:410 +0x174
  github.com/dop251/goja.(*baseJsFuncObject)._call()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/func.go:193 +0x8b3
  github.com/dop251/goja.(*baseJsFuncObject).call()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/func.go:203 +0x125
  github.com/dop251/goja.(*baseJsFuncObject).Call()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/func.go:156 +0x65
  github.com/dop251/goja.(*baseJsFuncObject).Call-fm()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/func.go:155 +0xae
  github.com/dop251/goja.AssertFunction.func1.2()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/runtime.go:2267 +0x136
  github.com/dop251/goja.(*vm).try()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/vm.go:547 +0x355
  github.com/dop251/goja.AssertFunction.func1()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/runtime.go:2266 +0x1c6
  go.k6.io/k6/js.(*VU).runFn()
      /home/mstoykov/work/k6io/k6/js/runner.go:786 +0x46d
  go.k6.io/k6/js.(*ActiveVU).RunOnce()
      /home/mstoykov/work/k6io/k6/js/runner.go:728 +0x887
  go.k6.io/k6/lib/executor.getIterationRunner.func1()
      /home/mstoykov/work/k6io/k6/lib/executor/helpers.go:145 +0x7d
  go.k6.io/k6/lib/executor.SharedIterations.Run.func4()
      /home/mstoykov/work/k6io/k6/lib/executor/shared_iterations.go:271 +0x5a7
  go.k6.io/k6/lib/executor.SharedIterations.Run·dwrap·31()
      /home/mstoykov/work/k6io/k6/lib/executor/shared_iterations.go:283 +0x58

Previous write at 0x00c000546840 by goroutine 43:
  github.com/dop251/goja.(*int8Array).set()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/typedarrays.go:198 +0x7e
  github.com/dop251/goja.(*typedArrayObject)._putIdx()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/typedarrays.go:520 +0x108
  github.com/dop251/goja.(*typedArrayObject).setOwnIdx()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/typedarrays.go:542 +0x4a
  github.com/dop251/goja.(*Object).setOwn()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/object.go:1403 +0x16c
  github.com/dop251/goja._setElemStrictP.exec()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/vm.go:1403 +0x235
      WARNING: DATA RACE
Write at 0x00c000546840 by goroutine 45:
  github.com/dop251/goja.(*int8Array).set()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/typedarrays.go:198 +0x7e
  github.com/dop251/goja.(*typedArrayObject)._putIdx()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/typedarrays.go:520 +0x108
  github.com/dop251/goja.(*typedArrayObject).setOwnIdx()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/typedarrays.go:542 +0x4a
  github.com/dop251/goja.(*Object).setOwn()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/object.go:1403 +0x16c
  github.com/dop251/goja._setElemStrictP.exec()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/vm.go:1403 +0x235
  github.com/dop251/goja.(*_setElemStrictP).exec()
      <autogenerated>:1 +0x44
  github.com/dop251/goja.(*vm).run()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/vm.go:410 +0x174
  github.com/dop251/goja.(*baseJsFuncObject)._call()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/func.go:193 +0x8b3
  github.com/dop251/goja.(*baseJsFuncObject).call()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/func.go:203 +0x125
  github.com/dop251/goja.(*baseJsFuncObject).Call()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/func.go:156 +0x65
  github.com/dop251/goja.(*baseJsFuncObject).Call-fm()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/func.go:155 +0xae
  github.com/dop251/goja.AssertFunction.func1.2()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/runtime.go:2267 +0x136
  github.com/dop251/goja.(*vm).try()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/vm.go:547 +0x355
  github.com/dop251/goja.AssertFunction.func1()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/runtime.go:2266 +0x1c6
  go.k6.io/k6/js.(*VU).runFn()
      /home/mstoykov/work/k6io/k6/js/runner.go:786 +0x46d
  go.k6.io/k6/js.(*ActiveVU).RunOnce()
      /home/mstoykov/work/k6io/k6/js/runner.go:728 +0x887
  go.k6.io/k6/lib/executor.getIterationRunner.func1()
      /home/mstoykov/work/k6io/k6/lib/executor/helpers.go:145 +0x7d
  go.k6.io/k6/lib/executor.SharedIterations.Run.func4()
      /home/mstoykov/work/k6io/k6/lib/executor/shared_iterations.go:271 +0x5a7
  go.k6.io/k6/lib/executor.SharedIterations.Run·dwrap·31()
      /home/mstoykov/work/k6io/k6/lib/executor/shared_iterations.go:283 +0x58

Previous write at 0x00c000546840 by goroutine 43:
  github.com/dop251/goja.(*int8Array).set()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/typedarrays.go:198 +0x7e
  github.com/dop251/goja.(*typedArrayObject)._putIdx()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/typedarrays.go:520 +0x108
  github.com/dop251/goja.(*typedArrayObject).setOwnIdx()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/typedarrays.go:542 +0x4a
  github.com/dop251/goja.(*Object).setOwn()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/object.go:1403 +0x16c
  github.com/dop251/goja._setElemStrictP.exec()
      /home/mstoykov/work/k6io/k6/vendor/github.com/dop251/goja/vm.go:1403 +0x235
```

I don't know how exactly this needs to be fixed

Reply:
Good catch! I must admit I can't think of any solution either at the moment. The main thing I see, is that your example builds upon the data being an `ArrayBuffer` hence, copiable to a typed array object such as `Int8Array`. Which leads me to think that I'd be better off with a dedicated object type under the hood, instead of relying on `ArrayBuffer` indeed. Cheers!

Reply:
As I mentioned earlier ... arguably the thing I imagined and is kind of discussed (IMO) in https://github.com/grafana/k6/issues/1931 is that we will have an immutable ArrayBuffer. 

Which at least I will expect to be put in Int8Array *but* it will throw an exception on trying to change its value, but not on reading it. So it will act exactly as one all the way through, except for writing data in it. (arguably the exception should only happen in strict mode but k6 is always in strict mode and I doubt we are changing that.)